### PR TITLE
WIP feat: Issue8 register ui funtionality

### LINKED ISSUE #8: Register a user
## Description
As a user,
I want to register on BIT,
So that I can sign up as mentor/mentee for programs I like

## Mocks
See the latest mock under https://github.com/anitab-org/bridge-in-tech-web/issues/2

## Linked Issue
[Create Registration API](https://github.com/anitab-org/bridge-in-tech-backend/issues/24
)
## Out of Scope
Handling duplicate registrations

### Update [Required]

- [ ] **Given** I am on registration page, 
       **When** I submit all details successfully,
       **Then** I see a message saying registration was successful. 

- [ ] **Given** I submit the registration form,
       **When** I receive an error from the backend service,
       **Then** I see a message saying there was an error 

- [ ] must have all the fields for Mentorship System user registration:
       [ ] name
       [ ] username
       [ ] email
       [ ] password(should be masked)
       [ ] terms and conditions
- [ ] must have available to be a `Mentor`, `Mentee` radio buttons
- [ ] must have `Signup` and `Login` buttons

- [ ] **Given** I am on the registration page,
        **When** I click on submit after filling the form,
        **Then** I see validation errors on the UI if any one of the following are an issue
        Email does not match xxx@xxx.com, password does not follow a combination of at least 1 
        digit, at least special character, at least uppercase letter, and length as 12 characters, confirm 
        password does not match password, any of the mandatory fields are missing, terms and 
        conditions is not checked. 
        
        Name: Min length = 2 , max length = 30
        Username: min = 5 max = 25
        Password: min = 8 max = 64
        Email: please match the pattern

For the "name" field to be valid, it may contain one or more character from:
- letter "a" to "z" and/or "A" to "Z", 
- any of the whitespace characters, and/or
- special character "-".

For the "email" field to be valid, it must have the following structure:
> the first section, which may contain one or more character from:
    - letter "a" to "z" and/or "A" to "Z",
    - number "0" to "9",
    - special character "_", ".", "+", and/or "-".
> followed by the "@" character,
> followed by the second section, which may contain one or more character from:
    - letter "a" to "z" and/or "A" to "Z",
    - number "0" to "9", and/or
    - special character "-".
> followed by the escaped character ".",
> followed by the third section, which may contain one or more character from:
    - letter "a" to "z" and/or "A" to "Z",
    - number "0" to "9",
    - special character "-" and/or ".".

For the "username" field to be valid, it may contain one or more character from:
    - letter "a" to "z" and/or "A" to "Z",
    - number "0" to "9", and/or
    - special character "-".

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentors.

## Estimation
2-3 days

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description

Created Register UI so user can register to BIT system.

Fixes #8 

### Type of Change:

- Code
- Quality Assurance
- User Interface

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade - npm dependencies)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
- running the app by executing command `npm start`. expected behaviours are achieved:
   * browser [shows Register](https://giphy.com/gifs/h53plVCRlOpr8zybCt/html5) form field when user click Register from top navbar from landing page
   * [fields validations](https://giphy.com/gifs/UWEbpYckB5HkRwdvaJ/html5) are matched to backend and basic UI interactions 

**Pending:**
- connecting Register UI with backend functionality

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by mtreacy002** [comment]
Update:
@anitab-org/mentorship-android-maintainers . I've pushed the code changes with the progress up to UI Register forms and validation fields features done. Please do start reviewing this changes while I'm continue working on connecting the backend. Test cases for this will be done in [issue#12](https://github.com/anitab-org/bridge-in-tech-web/issues/12) where I put my travis file so we can check if travis build pass for the tests.

File: src/Register.jsx [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        username: username || "",
+        email: email || "",
+        password: password || "",
+        passwordConfirm: passwordConfirm || "",
+        togglePassword: togglePassword || false
+    };
+  },
+  validationSchema: Yup.object().shape({
+    name: Yup.string()
+        .required("name is required")
```
Comment (foongminwong) [inline src/Register.jsx]: `.required("Name is required")`

**Comment by foongminwong** [comment]
The changes made in this PR were tested locally. Following are the results:

1. All possible responses are tested below:
   * **Navigate from Home to Register**  
     _Screenshot/gif/url_: 
test-pr-15-register-navigation

   * **Test Register Form**  
    _Screenshot/gif/url_: 
test-pr-15-register-form

     _Expected Result_:  Red warnings show up when (1) fields are left blank, (2) password too short, (3) confirm password doesn't match, (4) invalid email address when email is missing `@` or `.com`, etc. (5) username too short or with invalid characters
     _Actual Result_: Same as expected.

    * **Register a user using Register Form**
     _Screenshot/gif/url_:  
test-pr-15-register

     _Expected Result_:  User registration pending since the register functionality is WIP
     _Actual Result_:  Same as expected.

2. Additional comments:
   * Will test again when Register UI functionality is implemented.

3. OS Version: Windows 10

File: src/Register.jsx (hunk 1 above)
Comment (mtreacy002) [reply src/Register.jsx]: got it. will change it asap

**Comment by isabelcosta** [comment]
This is soooo awesome 👏 👏 @mtreacy002

**Comment by mtreacy002** [comment]
> This is soooo awesome 👏 👏 @mtreacy002

Thanks, @isabelcosta ❤️

**Comment by mtreacy002** [comment]
And Thank you @foongminwong for amazing testing 😍

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers .
So, I am now able to register user with only placing thee needed data on the form.

https://giphy.com/gifs/McUdtgipx2wKDVsTxV/html5

Animated GIF-source

Now I need to find out how to place other extra features (confirm password and toggle show/hide password field) and catching error responses. Any advice would be appreciated.

**Comment by meenakshi-dhanani** [comment]
If you think something can be committed, commit and push, so you can get early feedback

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers. I just pushed the latest update. You'll see many registersX.jsx files there. Please ignore it for now coz I'm still trying to workout my options. But for testing purposes, you can simply run the command `npm start` and the 2 BIT and MS servers. I placed the latest working code inside the route `/register`, so you don't have to do anything else. 😉

**Comment by mtreacy002** [comment]
PS: make sure you do the steps necessary for updating dependencies for you to test this changes. I would suggest the followings:
After you pull the PR, delete the existing package-lock.json (and node module if you have them in your own workspace) (**NOT THE package.json**). then reinstall dependencies using the `npm install` command. This will install whatever dependencies I have in the latest code change. You'll see new node module and package-lock.json got created as a result. Then go ahead with the testing 😉

**Comment by meenakshi-dhanani** [comment]
About the package.json and merge conflicts, there shouldn't be too many cases for a merge conflict considering you work alone and sequentially on the repo. Always as a rule, pull latest changes before starting to code. In case you have a merge conflict for packag-lock.json, you need not worry just rerun and accept your changes (Use VSCode to resolve them, please don't resolve by hand). So you'll need to establish a clean work around, if you pull in time and do a pull --rebase, considering you're the only contributor here, merge conflicts shouldn't be an issue

**Comment by meenakshi-dhanani** [comment]
Also, when you're experimenting with different ways of registering, you can branch out and try it. I would not prefer having options(tries) committed to code. Try keeping it clean, branch out into different registers if that helps. Also if you commit, there's no need to keep copies like this. You can always go back in history to get things. Henceforth, I'd prefer a trial branch for half-baked changes, instead of committing them to this PR. Because if this PR does get merged all those changes will come in the history and they don't seem clean.

**Comment by mtreacy002** [comment]
Ok, so I've made some changes since I've pushed last time till now. So, before next time I push again to the same PR branch, if you want me to pull rebase fist before I push, then the changes I've done just now will be gone. that won't be ideal

**Comment by meenakshi-dhanani** [comment]
Also, a question: Why can I see Login being committed? We're still on register in this PR

**Comment by mtreacy002** [comment]
> Also, a question: Why can I see Login being committed? We're still on register in this PR

because I have Login button on this form as part of UI. which if we want MVP, at least it should redirect to another page not just doing nothing, right?

**Comment by mtreacy002** [comment]
> About the package.json and merge conflicts, there shouldn't be too many cases for a merge conflict considering you work alone and sequentially on the repo. Always as a rule, pull latest changes before starting to code. In case you have a merge conflict for packag-lock.json, you need not worry just rerun and accept your changes (Use VSCode to resolve them, please don't resolve by hand). So you'll need to establish a clean work around, if you pull in time and do a pull --rebase, considering you're the only contributor here, merge conflicts shouldn't be an issue

Also, accepting changes on merge conflict when I pull after I add new dependencies means that I agree to use the old dependencies and remove my latest one isn't it? But I want to have the latest change. So far I VS code, I select either accepting incoming changes or accepting current changes per each conflict, not as whole coz I want to be sure that I'm selecting the right codes to move forward (plus I'm not sure if we can do that (accepting all changes in one click) in VS Code 🤔 )

**Comment by mtreacy002** [comment]
> Also, when you're experimenting with different ways of registering, you can branch out and try it. I would not prefer having options(tries) committed to code. Try keeping it clean, branch out into different registers if that helps. Also if you commit, there's no need to keep copies like this. You can always go back in history to get things. Henceforth, I'd prefer a trial branch for half-baked changes, instead of committing them to this PR. Because if this PR does get merged all those changes will come in the history and they don't seem clean.

So, you're ok for me to push to the same PR with different branches? Or is that mean I have to open new PR and link them to the same issue?

**Comment by mtreacy002** [comment]
Isn't it possible to delete the non-used trials (register files) later once we confirm which one works (with git squash and push -f) to show only the latest files on github?

**Comment by mtreacy002** [comment]
This PR is now closed and will be replaced with another one.

**Pull request closed without merging**